fastestimator/estimator.py

```diff
@@ -64,92 +64,104 @@ def __init__(self,
-                             global_step=0,
-                             batch_size=None,
-                             num_devices=get_num_devices(),
-                             log_steps=self.log_steps,
-                             total_epochs=self.epochs,
-                             total_steps=None,
-                             epoch_idx=0,
-                             batch_idx=0)
+        system = System(mode="train",
+                        global_step=0,
+                        num_devices=get_num_devices(),
+                        log_steps=self.log_steps,
+                        total_epochs=self.epochs,
+                        epoch_idx=0,
+                        batch_idx=0)
+        system.network = self.network
         for trace in self.traces:
-            trace.system = self.system
+            trace.system = system
+        return system
 
     def _prepare_traces(self):
-        loss_keys = self._initialize_loss_keys()
-        if self.traces is None:
-            self.traces = []
-        self.traces = to_list(self.traces)
-        self.monitor_names = set(filter(None, to_list(self.monitor_names))).union(loss_keys)
-        self.traces.insert(0, TrainEssential(monitor_names=self.monitor_names))
-        if self.do_eval:
-            self.traces.append(EvalEssential(loss_keys=loss_keys))
-        for trace in self.traces:
-            self.trace_inputs = self.trace_inputs.union(set(filter(None, to_list(trace.inputs))))
-            self.monitor_names = self.monitor_names.union(set(filter(None, to_list(trace.log_names))))
-        self.traces.append(Logger(log_names=self.monitor_names, loss_names=loss_keys))
-
-    def _initialize_loss_keys(self) -> Set[str]:
-        loss_keys = set()
-        for op in self.network.ops:
-            if isinstance(op, UpdateOp):
-                loss_keys = loss_keys.union(set(to_list(op.inputs)))
-        return loss_keys
+        self.trace_inputs = dict()
+        traces = [trace for trace in self.traces]
+        loss_keys = self.network.get_loss_keys()
+        monitor_names = set(filter(None, to_list(self.monitor_names))).union(loss_keys)
+        traces.insert(0, TrainEssential(monitor_names=monitor_names))
+        modes = self.pipeline.get_modes() - {"test"}
+        if "eval" in modes:
+            traces.append(EvalEssential(loss_keys=loss_keys))
+        for mode in modes:
+            trace_inputs = set()
+            for trace in traces:
+                if trace.mode is None or mode in to_list(trace.mode):
+                    trace_inputs.update(filter(None, to_list(trace.inputs)))
+                    monitor_names.update(filter(None, to_list(trace.log_names)))
+            self.trace_inputs[mode] = trace_inputs
+        traces.append(Logger(log_names=monitor_names, loss_names=loss_keys))
+        return traces
 
     def _start(self):
         self._run_traces_on_begin()
         for self.system.epoch_idx in range(self.epochs):
```

Comment:
should self.epochs be self.system.epochs ?

Reply:
they are the same number.   but I did change the variable name from self.system.total_epochs to self.system.epochs

Reply:
I guess the point of this question was more: why are we storing self.epochs instead of exclusively relying on self.system.epochs?

Reply:
the system is instantiated after we call the fit, but we need self.epochs to store the init argument.  there's absolutely no differences between either case

Reply:
storing same info in multiple places can give rise to inconsistent states. better to have single source of truth when possible. maybe not possible here, but we could perhaps partially instantiate system on estimator creation

Reply:
The system has to re-initiate itself or reset itself after every fit call.  - this is critical for notebook usage.  Therefore we might as well instantiate new system for every fit.